iocore/net/P_SNIActionPerformer.h

```diff
@@ -89,12 +90,73 @@ class HTTP2BufferWaterMark : public ActionItem
 
 class TunnelDestination : public ActionItem
 {
+  // ID of the configured variable. This will be used to know which function
+  // should be called when processing the tunnel destination.
+  enum OpId : int32_t {
+    DEFAULT = -1,                 // No specific variable set.
+    MATCH_GROUPS,                 // Deal with configured groups.
+    MAP_WITH_RECV_PORT,           // Use port from inbound local
+    MAP_WITH_PROXY_PROTOCOL_PORT, // Use port from the proxy protocol
+    MAX                           // Always at the end and do not change the value of the above items.
+  };
+  const std::string MAP_WITH_RECV_PORT_STR           = "{inbound_local_port}";
+  const std::string MAP_WITH_PROXY_PROTOCOL_PORT_STR = "{proxy_protocol_port}";
+
 public:
   TunnelDestination(const std::string_view &dest, SNIRoutingType type, YamlSNIConfig::TunnelPreWarm prewarm,
                     const std::vector<int> &alpn)
     : destination(dest), type(type), tunnel_prewarm(prewarm), alpn_ids(alpn)
   {
-    need_fix = (destination.find_first_of('$') != std::string::npos);
+    std::size_t var_start_pos{0}; // keep the starting position of the variable.
+    if (destination.find_first_of('$') != std::string::npos) {
+      fnArrIndex = OpId::MATCH_GROUPS;
+    } else if (var_start_pos = destination.find(MAP_WITH_RECV_PORT_STR); var_start_pos != std::string::npos) {
+      fnArrIndex = OpId::MAP_WITH_RECV_PORT;
+    } else if (var_start_pos = destination.find(MAP_WITH_PROXY_PROTOCOL_PORT_STR); var_start_pos != std::string::npos) {
+      fnArrIndex = OpId::MAP_WITH_PROXY_PROTOCOL_PORT;
+    }
+
+    // Replace wildcards with matched groups.
+    fix_destination[OpId::MATCH_GROUPS] = [&](std::string_view destination, const Context &ctx, SSLNetVConnection *,
+                                              bool &port_is_dynamic) -> std::string {
+      port_is_dynamic = false;
+      if ((destination.find_first_of('$') != std::string::npos) && ctx._fqdn_wildcard_captured_groups) {
+        const auto &fixed_dst = replace_match_groups(destination, *ctx._fqdn_wildcard_captured_groups, port_is_dynamic);
+        return fixed_dst;
```

Comment:
Isn't `fixed_dst` a reference? Is it safe to return?

Reply:
No, it is not safe to return the reference. Thanks for catching this.

Reply:
> Isn't `fixed_dst` a reference? Is it safe to return?

It should be safe, the function returns by value, the caller should get this stored safely.